File: api/awssig4.py
import hashlib
import hmac
import datetime

region = 'us-west-2'
service = 's3'

canonical_uri = '/'
canonical_querystring = ''
canonical_headers = 'host:s3.amazonaws.com\nx-amz-date:20210717T000000Z\n'
payload = ''


def sign(key, msg):
    return hmac.new(key, msg.encode('utf-8'), hashlib.sha256).digest()


def getSignatureKey(key, dateStamp, regionName, serviceName):
    kDate = sign(('AWS4' + key).encode('utf-8'), dateStamp)
    kRegion = sign(kDate, regionName)
    kService = sign(kRegion, serviceName)
    kSigning = sign(kService, 'aws4_request')
    return kSigning


def calculate_aws4_signature(api_key, api_secret, http_method, canonical_headers, timestamp):
    # Step 1: Create a date for headers and the credential string
    t = datetime.datetime.strptime(timestamp, '%Y%m%dT%H%M%SZ')
    amz_date = t.strftime('%Y%m%dT%H%M%SZ')  # Format: YYYYMMDD'T'HHMMSS'Z'
    # Date w/o time, used in credential scope
    date_stamp = t.strftime('%Y%m%d')

    # Step 2: Create canonical request
    payload_hash = hashlib.sha256(payload.encode('utf-8')).hexdigest()
    canonical_request = f"{http_method}\n{canonical_uri}\n{canonical_querystring}\n{canonical_headers}\n{payload_hash}"

    # Step 3: Create the string to sign
    algorithm = 'AWS4-HMAC-SHA256'
    credential_scope = f"{date_stamp}/{region}/{service}/aws4_request"
    string_to_sign = f"{algorithm}\n{amz_date}\n{credential_scope}\n{hashlib.sha256(canonical_request.encode('utf-8')).hexdigest()}"

    # Step 4: Calculate the signature
    signing_key = getSignatureKey(api_secret, date_stamp, region, service)
    signature = hmac.new(signing_key, string_to_sign.encode(
        'utf-8'), hashlib.sha256).hexdigest()

    # Step 5: Add signing information to the request
    authorization_header = f"{algorithm} Credential={api_key}/{credential_scope}, SignedHeaders=host;x-amz-date, Signature={signature}"

    return {
        "x-amz-date": amz_date,
        "Authorization": authorization_header
    }


# Example usage
if __name__ == 'main':
    api_key = '55305bd623f10f22b8f3427916da8b0a'
    api_secret = '28b5b016e14c839c85cd736e30d7ed0a'
    timestamp = '20210717T000000Z'  # Example timestamp
    http_method = 'GET'
    headers = calculate_aws4_signature(
        api_key, api_secret, http_method, timestamp)
    print(headers)

File: frontend/src/api/awssig4.js
// Include the CryptoJS library via CDN in your HTML file:
// <script src="https://cdnjs.cloudflare.com/ajax/libs/crypto-js/4.0.0/crypto-js.min.js"></script>

const region = "us-west-2";
const service = "s3";

const canonicalUri = "/";
const canonicalQuerystring = "";
const canonicalHeaders = "host:s3.amazonaws.com\nx-amz-date:20210717T000000Z\n";
const payload = "";

const sign = (key, msg) =>
  CryptoJS.HmacSHA256(msg, key);

const getSignatureKey = (key, dateStamp, regionName, serviceName) => {
  const kDate = sign(`AWS4${key}`, dateStamp);
  const kRegion = sign(kDate, regionName);
  const kService = sign(kRegion, serviceName);
  const kSigning = sign(kService, "aws4_request");
  return kSigning;
};

const calculateAws4Signature = (apiKey, apiSecret, httpMethod) => {
  // Step 1: Create a date for headers and the credential string
  const t = new Date();
  const amzDate = t.toISOString().replace(/[:-]|\.\d{3}/g, ""); // Format: YYYYMMDD'T'HHMMSS'Z'
  const dateStamp = amzDate.substr(0, 8); // Date w/o time, used in credential scope

  // Step 2: Create canonical request
  const payloadHash = CryptoJS.SHA256(payload).toString(CryptoJS.enc.Hex);
  const canonicalRequest = `${httpMethod}\n${canonicalUri}\n${canonicalQuerystring}\n${canonicalHeaders}\n${payloadHash}`;

  // Step 3: Create the string to sign
  const algorithm = "AWS4-HMAC-SHA256";
  const credentialScope = `${dateStamp}/${region}/${service}/aws4_request`;
  const stringToSign = `${algorithm}\n${amzDate}\n${credentialScope}\n${CryptoJS.SHA256(canonicalRequest).toString(CryptoJS.enc.Hex)}`;

  // Step 4: Calculate the signature
  const signingKey = getSignatureKey(apiSecret, dateStamp, region, service);
  const signature = CryptoJS.HmacSHA256(stringToSign, signingKey).toString(CryptoJS.enc.Hex);

  // Step 5: Add signing information to the request
  const authorizationHeader = `${algorithm} Credential=${apiKey}/${credentialScope}, SignedHeaders=host;x-amz-date, Signature=${signature}`;

  return {
    "x-amz-date": amzDate,
    Authorization: authorizationHeader,
  };
};

// Example usage
/*
const apiKey = "55305bd623f10f22b8f3427916da8b0a";
const apiSecret = "28b5b016e14c839c85cd736e30d7ed0a";
const httpMethod = "GET";
const headers = calculateAws4Signature(apiKey, apiSecret, httpMethod);
console.log(headers);
*/

export default calculateAws4Signature;
